nucypher/policy/orders.py

```diff
@@ -38,310 +38,261 @@
+        self.encrypted_kfrag = encrypted_kfrag
+        self.capsules = capsules
 
-    def sanitize(self):
-        for task in self.tasks.values():
-            task.cfrag = CFRAG_NOT_RETAINED
+    def __bytes__(self):
+        return (bytes(self.hrac) +
+                bytes(self._alice_verifying_key) +
+                bytes(self._bob_verifying_key) +
+                bytes(self._publisher_verifying_key) +
+                VariableLengthBytestring(self.encrypted_kfrag) +
+                b''.join(bytes(capsule) for capsule in self.capsules)
+                )
 
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        hrac, alice_vk, bob_vk, publisher_vk, ekfrag, remainder = cls._splitter(data, return_remainder=True)
+        capsules = capsule_splitter.repeat(remainder)
+        return cls(hrac, alice_vk, bob_vk, publisher_vk, ekfrag, capsules)
 
-class WorkOrderHistory:
+    def alice(self) -> 'Alice':
+        from nucypher.characters.lawful import Alice
+        return Alice.from_public_keys(verifying_key=self._alice_verifying_key)
 
-    def __init__(self) -> None:
-        self.by_ursula = {}  # type: dict
-        self._latest_replete = {}
+    def bob(self) -> 'Bob':
+        from nucypher.characters.lawful import Bob
+        return Bob.from_public_keys(verifying_key=self._bob_verifying_key)
 
-    def __contains__(self, item):
-        assert False
+    def publisher(self) -> 'Alice':
+        from nucypher.characters.lawful import Alice
+        return Alice.from_public_keys(verifying_key=self._publisher_verifying_key)
 
-    def __getitem__(self, item):
-        return self.by_ursula[item]
 
-    def __setitem__(self, key, value):
-        assert False
+class ReencryptionResponse:
+    """
+    A response from Ursula with reencrypted capsule frags.
+    """
 
-    def __len__(self):
-        return sum(len(work_orders) for work_orders in self.by_ursula.values())
+    @classmethod
+    def construct_by_ursula(cls,
+                            capsules: List[Capsule],
+                            cfrags: List[VerifiedCapsuleFrag],
+                            stamp: SignatureStamp,
+                            ) -> 'ReencryptionResponse':
 
-    @property
-    def ursulas(self):
-        return self.by_ursula.keys()
+        # un-verify
+        cfrags = [CapsuleFrag.from_bytes(bytes(cfrag)) for cfrag in cfrags]
+
+        capsules_bytes = b''.join(bytes(capsule) for capsule in capsules)
+        cfrags_bytes = b''.join(bytes(cfrag) for cfrag in cfrags)
+        signature = stamp(capsules_bytes + cfrags_bytes)
+        return cls(cfrags, signature)
+
+    def __init__(self, cfrags: List[CapsuleFrag], signature: Signature):
+        self.cfrags = cfrags
+        self.signature = signature
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        signature, cfrags_bytes = signature_splitter(data, return_remainder=True)
+        cfrags = cfrag_splitter.repeat(cfrags_bytes)
```

Comment:
Do we need to check whether `cfrags_bytes`  is not None first? Similar to what we did for `RetrievalKit`?

Reply:
A response with zero cfrags makes no sense because we won't ever make a request with zero capsules. But I guess it can be checked for a more informative error.

Reply:
Fixed.